_posts/work/2018-01-30-hdf-in-the-cloud.md

```diff
@@ -0,0 +1,311 @@
+to more efficiently query their cloud-hosted datasets.
+
+
+New Formats for Scientific Data
+-------------------------------
+
+Alternatively, we can move on from the HDF file format,
+and invent a new data storage specification that fits cloud storage
+(or other storage)
+more cleanly without worrying about supporting the legacy layout of existing HDF files.
+
+This has already been going on, informally, for years.
+Most often we see people break large arrays into blocks,
+store each block as a separate object in the cloud object store with a suggestive name,
+and store a metadata file describing how the blocks relate to each other.
+This looks something like the following:
+
+```
+/metadata.json
+/0.0.0.dat
+/0.0.1.dat
+/0.0.2.dat
+ ...
+/10.10.8.dat
+/10.10.9.dat
+/10.10.10.dat
+```
+
+There are many variants:
+
+- They might extend this to have groups or sub-arrays in sub-directories.
+- They might choose to compress the individual blocks in the `.dat` files or not.
+- They might choose different encoding schemes for the metadata and the binary blobs.
+
+But generally most array people on the cloud do something like this with their research data,
+and they've been doing it for years.
+It works efficiently,
+is easy to understand and manage,
+and transfers well to any cloud platform,
+onto a local file system,
+or even into a standalone zip file or small database.
+
+There are two groups that have done this in a more mature way,
+defining both modular standalone libraries to manage their data,
+as well as proper specification documents that inform others how to interpret this data in a long-term stable way.
+
+-  [Zarr](http://zarr.readthedocs.io/en/stable/)
+-  [N5](https://github.com/saalfeldlab/n5)
+
+These are both well maintained and well designed libraries (by my judgment),
+in Python and Java respectively.
+They offer layouts like the layout above, although with more sophistication.
+Entertainingly [their](https://github.com/saalfeldlab/n5#filesystem-specification-version-100) [specs](http://zarr.readthedocs.io/en/stable/spec/v2.html) are similar enough that another library,
+[Z5](https://github.com/constantinpape/z5),
+built a cross-compatible parser for each in C++.
+This unintended uniformity is a good sign.
+It means that both developer groups were avoiding creativity,
+and have converged on a sensible common solution.
+I encourage you to read the [Zarr Spec](http://zarr.readthedocs.io/en/stable/spec/v2.html) in particular.
+
+However, technical merits are not alone sufficient to justify a shift in data format,
+especially for archival datasets of record that we're discussing.
+The institutions in charge of this data and have multi-decade horizons and so move slowly.
+For them, moving off of the historically community standard would be major shift.
+
+And so we need to answer a couple of difficult questions:
+
+1.  How hard is it to make HDF efficient in the cloud?
+2.  How hard is it to shift the community to a new standard?
```

Comment:
I would close with something about the urgency of answering these questions *before* NASA's [cumulus project](https://earthdata.nasa.gov/eosdis-cumulus-project) starts pushing PB of data.

Reply:
When cumulus start pushing data?

Reply:
https://earthdata.nasa.gov/cmr-and-esdc-in-cloud
> From its current cumulative archive size of almost 22 petabytes (PB), the volume of data in the EOSDIS archive is expected to grow to almost 247 PB by 2025, according to estimates by NASA’s Earth Science Data Systems (ESDS) Program. Over the next five years, the daily ingest rate of data into the EOSDIS archive is expected to reach 114 terabytes (TB) per day, with the upcoming joint NASA/Indian Space Research Organization Synthetic Aperture Radar (NISAR) mission (scheduled for launch by 2021) contributing an estimated 86 TB per day of data when operational.